controller/sharding/sharding.go

```diff
@@ -158,6 +162,59 @@ func RoundRobinDistributionFunction(clusters clusterAccessor, replicas int) Dist
 	}
 }
 
+// ConsistentHashingWithBoundedLoadsDistributionFunction returns a DistributionFunction using an almost homogeneous distribution algorithm:
+// for a given cluster the function will return the shard number based on a consistent hashing with bounded loads algorithm.
+// This function ensures an almost homogenous distribution: each shards got assigned the fairly similar number of
+// clusters +/-10% , but with it is resilient to sharding and/or number of clusters changes.
+func ConsistentHashingWithBoundedLoadsDistributionFunction(clusters clusterAccessor, replicas int) DistributionFunction {
+	return func(c *v1alpha1.Cluster) int {
+		if replicas > 0 {
+			if c == nil { // in-cluster does not necessarly have a secret assigned. So we are receiving a nil cluster here.
+				return 0
+			}
+
+			// if Shard is manually set and the assigned value is lower than the number of replicas,
+			// then its value is returned otherwise it is the default calculated value
+			if c.Shard != nil && int(*c.Shard) < replicas {
+				return int(*c.Shard)
+			} else {
+				// if the cluster is not in the clusters list anymore, we should unassign it from any shard, so we
+				// return the reserved value of -1
+				if !slices.Contains(clusters(), c) {
+					log.Warnf("Cluster with id=%s not found in cluster map.", c.ID)
+					return -1
+				}
+				consistentHashing := createConsistentHashingWithBoundLoads(replicas)
+				clusterIndex, err := consistentHashing.Get(c.ID)
+				if err != nil {
+					log.Warnf("Cluster with id=%s not found in cluster map.", c.ID)
+					return -1
+				}
+				shard, err := strconv.Atoi(clusterIndex)
+				if err != nil {
+					log.Errorf("Consistent Hashing was supposed to return a shard index but it returned %d", err)
+					return -1
+				}
+				log.Debugf("Cluster with id=%s will be processed by shard %d", c.ID, shard)
+				return shard
+			}
+		}
+		log.Warnf("The number of replicas (%d) is lower than 1", replicas)
+		return -1
+	}
+}
+
+func createConsistentHashingWithBoundLoads(replicas int) *consistent.Consistent {
+	consistentHashing := consistent.New()
+	// Adding a shard with id "-1" as a reserved value for clusters that does not have an assigned shard
+	// this happens for clusters that are removed for the clusters list
+	//consistentHashing.Add("-1")
+	for i := 0; i < replicas; i++ {
+		consistentHashing.Add(strconv.Itoa(i))
+	}
+	return consistentHashing
+}
```

Comment:
Modifications to the createConsistentHashsingWithBoundLoads function. This will create the map for cluster to shard. Note that the list must be consistent across all shards so that is why the cluster list must be sorted before going through the consistent hash algorithm:
```
func createConsistentHashingWithBoundLoads(replicas int, getCluster clusterAccessor, getApp appAccessor) map[string]int {
	clusters := getSortedClustersList(getCluster)
	appDistribution := getAppDistribution(getCluster, getApp)
	shardIndexedByCluster := make(map[string]int)	
	appsIndexedByShard := make(map[string]int64)	
	consistentHashing := consistent.New()
	// Adding a shard with id "-1" as a reserved value for clusters that does not have an assigned shard
	// this happens for clusters that are removed for the clusters list
	//consistentHashing.Add("-1")
	for i := 0; i < replicas; i++ {
		shard := strconv.Itoa(i)
		consistentHashing.Add(shard)
		appsIndexedByShard[shard] = 0
	}

	for _, c := range clusters {
		clusterIndex, err := consistentHashing.GetLeast(c.ID)
		if err != nil {
			log.Warnf("Cluster with id=%s not found in cluster map.", c.ID)
		}
		shardIndexedByCluster[c.ID], err = strconv.Atoi(clusterIndex)
		if err != nil {
			log.Errorf("Consistent Hashing was supposed to return a shard index but it returned %d", err)
		}
		numApps, ok := appDistribution[c.Server]
		if !ok {
			numApps = 0
		} 
		appsIndexedByShard[clusterIndex] += numApps
		consistentHashing.UpdateLoad(clusterIndex, appsIndexedByShard[clusterIndex])
	}

	return shardIndexedByCluster
}
```

Reply:
Adding the sort and appsIndexByShard.